# Homework 4

## Graphical Parameters and Model Structure

In the previous homework, you performed queries on a graphical model of possible murders and murder weapons. Now, you will estimate model parameters and structure using data.   

As a reminder, the joint probability distribution is:

$$p(B,C,BW,CW,M)$$     

where the letters indicate the following variables;   
$B = $ butler committed the crime, {not murderer, murderer},   
$C = $ cook committed the crime, {not murderer, murderer},    
$BW = $ choice of weapon, {knife, knife with poison}, conditional on butler,  
$CW = $ choice of weapon, {knife, knife with poison}, conditional on cook,   
$M = $ murderer {butler or cook, third party alone, combination of butler or cook and third party}.    

Keeping in mind it is possible the cook, the butler and the third party could be guilty, and that participation in the crime in any capacity constitutes guilt, the distribution can be factorized in the following manner:

$$p(B,C,BW,CW,M) = p(B)\ p(C)\ p(BW\ |\ B)\ p(CW\ |\ C)\ p(M\ |\ BK,CW, C, W)$$  

A graph of the model is shown below. 

<img src="MurderDirected.JPG" alt="Drawing" style="width:600px; height:300px"/>
<center> **DAG for murder evidence** </center>

Notice that the skeleton of this graph does not have a tree structure. This fact will limit how well estimation algorithms will work, particularly for graph structure. Keep this fact in mind as you proceed. 

As a first step execute the code in the below to simulate the 20 cases for the dataset. Examine the code to see the CPD tables for this simulation. 

In [2]:
## Simulate the binary tables
import numpy as np
import numpy.random as nr
import pandas as pd

def sim_bernoulli(p, n = 25):
    """
    Function to compute the vectors with probabilities for each 
    condition (input value) of the dependent variable using the Bernoulli
    distribution. 
    
    The arguments are:
    p - a vector of probabilites of success for each case.
    n - The numer of realizations. 
    """
    temp = np.zeros(shape = (len(p), n))
    for i in range(len(p)): 
        temp[i,:] = nr.binomial(1, p[i], n)
    return(temp)

def selec_dist_1(sims, var, lg):
    """
    Function to integrate the conditional probabilities for
    each of the cases of the parent variable. 
    
    The arguments are:
    sims - the array of simulated realizations with one row for each state of the
           parent variable. 
    var - the vector of values of parent variable used to select the value from the 
          sims array.
    lg - vector of states of possible states of the parent variable. These must be
         in the same order as for the sims array. 
    """
    out = sims[0,:] # Copy of values for first parent state
    var = np.array(var).ravel()
    for i in range(1, sims.shape[0]): # loop over other parent states
        out = [x if u == lg[i] else y for x,y,u in zip(sims[i,:], out, var)]
    return([int(x) for x in out])

def set_class_2(x):
    """
    Function to flatten the array produced by the numpy.random.multinoulli function. 
    The function tests which binary value of the array of output states is true
    and substitutes an integer for that state. This function only works for up to three
    output states. 
    
    Argument:
    x - The array produced by the numpy.random.multinoulli function. 
    """
    out = []
    for i,j in enumerate(x):
        if j[0] == 1: out.append(0)
        elif j[1] == 1: out.append(1)
        else: out.append(2)       
    return(out)   

def sim_multinoulli(p, n = 25):
    """
    Function to compute the vectors with probabilities for each 
    condition (input value) of the dependent variable using the multinoulli
    distribution. 
    
    The arguments are:
    p - an array of probabilites of success for each possible combination
        of states of the parent variables. Each row in the array are the 
        probabilities for each state of the multinoulli distribution for 
        that combination of parent values.
    n - The numer of realizations. 
    """
    temp = np.zeros(shape = (p.shape[0], n))
    for i in range(p.shape[0]):  
        ps = p[i,:]
        mutlis = nr.multinomial(1, ps, n) 
        temp[i,:] = set_class_2(mutlis)
    return(temp)

def selec_dist_4(sims, var1, var2, var3, var4, lg1, lg2, lg3, lg4):
    """
    Function to integrate the conditional probabilities for
    each of the cases of four parent variables. 
    
    The arguments are:
    sims - the array of simulated realizations with one row for each state of the
           union of the parent variables. 
    var1 - the vector of values of first parent variable used to select the value from the 
          sims array.
    var2 - the vector of values of second parent variable used to select the value from the 
          sims array.
    var3 - the vector of values of third parent variable used to select the value from the 
          sims array.
    var4 - the vector of values of fourth parent variable used to select the value from the 
          sims array.
    lg1 - vector of states of possible states of the first parent variable. These must be
         in the same order as for the sims array. 
    lg2 - vector of states of possible states of the second parent variable. These must be
         in the same order as for the sims array. 
    lg3 - vector of states of possible states of the third parent variable. These must be
         in the same order as for the sims array. 
    lg4 - vector of states of possible states of the fourth parent variable. These must be
         in the same order as for the sims array. 
    """
    out = sims[0,:] # Copy values for first combination of states for parent variables
    
    ## make sure the parent variables are 1-d numpy arrays.
    var1 = np.array(var1).ravel() 
    var2 = np.array(var2).ravel()
    var3 = np.array(var3).ravel() 
    var4 = np.array(var4).ravel()
    for i in range(1, sims.shape[0]): # Loop over all combination of states of the parent variables
        out = [x if t == lg1[i] and u == lg2[i] and v == lg3[i] and w == lg4[i] else y for x,y,t,u,v,w in 
              zip(sims[i,:], out, var1, var2, var3, var4)]                         
    return([int(x) for x in out])


## set the sample size
nsamp = 20

## First the conditionally independent variables
nr.seed(22234)
B_samps = pd.DataFrame(nr.binomial(1, 0.8, nsamp), columns = ['B'])
nr.seed(2355)
C_samps = pd.DataFrame(nr.binomial(1, 0.3, nsamp), columns = ['C'])

## Two variables conditionally depenent on one other
probs = [0.5, 0.1]
nr.seed(2134)
bern = sim_bernoulli(probs, nsamp)
BW_samps = pd.DataFrame(selec_dist_1(bern, B_samps, [0,1]), columns = ['BW'])

probs = [0.5, 0.7]
nr.seed(22234)
bern = sim_bernoulli(probs)
CW_samps = pd.DataFrame(selec_dist_1(bern, C_samps, [0,1]), columns = ['CW'])


probs = np.array([[0.0, 0.7, 0.5,  0.7, 0.0, 0.6, 0.5,  0.6, 0.0, 0.7, 0.5,  0.7, 0.0, 0.6, 0.5,  0.6],
                           [1.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0],
                           [0.0, 0.1, 0.4,  0.2, 0.0, 0.1, 0.3,  0.2, 0.0, 0.1, 0.4,  0.2, 0.0, 0.1, 0.3,  0.2]]).transpose()

nr.seed(2334)
sims = sim_multinoulli(probs, nsamp)

C_lg = [0,0,1,1,0,0,1,1,0,0,1,1,0,0,1,1]
B_lg = [0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1]
BW_lg = [0,0,0,0,1,1,1,1,0,0,0,0,1,1,1,1]
CW_lg = [0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1]
M_samps = pd.DataFrame(selec_dist_4(sims, B_samps, C_samps, BW_samps, CW_samps, B_lg, C_lg, BW_lg, CW_lg), columns = ['M'])



## Now concatenate the columns into one data frame
dats = pd.concat([B_samps, C_samps, BW_samps, CW_samps, M_samps], axis = 1)
print(dats.shape)
dats

(20, 5)


B  C  BW  CW  M
0   1  0   0   1  0
1   1  0   0   1  0
2   1  1   0   1  2
3   1  0   0   0  0
4   1  1   0   0  0
5   0  1   1   0  0
6   1  1   0   1  2
7   1  0   0   0  0
8   1  0   0   0  0
9   0  0   1   1  1
10  0  0   0   1  1
11  1  0   0   1  0
12  0  0   1   1  1
13  1  0   0   0  0
14  1  0   0   0  0
15  1  1   0   0  0
16  1  1   0   1  2
17  1  0   0   0  0
18  1  0   0   0  0
19  0  0   1   1  1

## Part 1: Parameter Estimation

With the dataset generated you will now estimate the parameters of the graphical model using both maximum likelihood and Bayesian methods. 

As a first step execute the code in the cell below to load the packages you will need.

In [3]:
from pgmpy.models import BayesianModel
from pgmpy.estimators import MaximumLikelihoodEstimator, BayesianEstimator
from pgmpy.inference import BeliefPropagation
from pgmpy.estimators import HillClimbSearch, BicScore, K2Score, StructureScore

Now, create and execute the code in the cell below to estimate and display the parameters of the CPDs using the graph structure previously shown.

In [4]:

model = BayesianModel([('B','BW'),('C','CW'),('B','M'),('C','M'),('BW','M'),("CW",'M')])
model.fit(dats, estimator=MaximumLikelihoodEstimator)
for cpd in model.get_cpds():
    print("CPD of {variable}:".format(variable=cpd.variable))
    print(cpd)

CPD of B:
╒══════╤══════╕
│ B(0) │ 0.25 │
├──────┼──────┤
│ B(1) │ 0.75 │
╘══════╧══════╛
CPD of BW:
╒═══════╤══════╤══════╕
│ B     │ B(0) │ B(1) │
├───────┼──────┼──────┤
│ BW(0) │ 0.5  │ 0.5  │
├───────┼──────┼──────┤
│ BW(1) │ 0.5  │ 0.5  │
╘═══════╧══════╧══════╛
CPD of C:
╒══════╤═════╕
│ C(0) │ 0.7 │
├──────┼─────┤
│ C(1) │ 0.3 │
╘══════╧═════╛
CPD of CW:
╒═══════╤══════╤══════╕
│ C     │ C(0) │ C(1) │
├───────┼──────┼──────┤
│ CW(0) │ 0.5  │ 0.5  │
├───────┼──────┼──────┤
│ CW(1) │ 0.5  │ 0.5  │
╘═══════╧══════╧══════╛
CPD of M:
╒══════╤════════════════════╤═══════╤════════════════════╤════════════════════╤════════════════════╤═══════╤═══════╤════════════════════╤═══════╤═══════╤═══════╤═══════╤════════════════════╤════════════════════╤════════════════════╤════════════════════╕
│ B    │ B(0)               │ B(0)  │ B(0)               │ B(0)               │ B(0)               │ B(0)  │ B(0)  │ B(0)               │ B(1)  │ B(1)  │ B(1)  │ B(1)  │ B(1)               │ B(1)        

Examine these results and answer the following questions:
1. How many parameters are there in the CPD tables?
2. Given the number of cases, is this MLE problem underdetermined? 

ANS 1: We have 64 parameters in the CPD tables (N-1 = 63)

ANS 2: In this case, it is likely MLE problem of underdetermination. Because too small data set.

Next, you will estimate the CPD parameters using the Bayesian estimator. For this first estimate use the following moderately weak and uniform prior distributions (pseudo counts):

- C: {4,4}
- B: {4,4}
- CW: {4,4}
- BW: {4,4}
- M: {2,2,2}

In the cell below create and execute the code to perform Bayes estimation and display the CPD parameters. 

In [5]:
bayes_model = BayesianModel([('B','BW'),('C','CW'),('B','M'),('C','M'),('BW','M'),("CW",'M')])

pseudo_counts = {'C': [4,4], 'B': [4,4],'CW': [4,4],'BW': [4,4],'M': [2,2,2]}
bayes_model.fit(dats, estimator=BayesianEstimator, prior_type='dirichlet', pseudo_counts=pseudo_counts)

for cpd in bayes_model.get_cpds():
    print("CPD of {variable}:".format(variable=cpd.variable))
    print(cpd)

CPD of B:
╒══════╤══════════╕
│ B(0) │ 0.321429 │
├──────┼──────────┤
│ B(1) │ 0.678571 │
╘══════╧══════════╛
CPD of BW:
╒═══════╤══════╤══════╕
│ B     │ B(0) │ B(1) │
├───────┼──────┼──────┤
│ BW(0) │ 0.5  │ 0.5  │
├───────┼──────┼──────┤
│ BW(1) │ 0.5  │ 0.5  │
╘═══════╧══════╧══════╛
CPD of C:
╒══════╤══════════╕
│ C(0) │ 0.642857 │
├──────┼──────────┤
│ C(1) │ 0.357143 │
╘══════╧══════════╛
CPD of CW:
╒═══════╤══════╤══════╕
│ C     │ C(0) │ C(1) │
├───────┼──────┼──────┤
│ CW(0) │ 0.5  │ 0.5  │
├───────┼──────┼──────┤
│ CW(1) │ 0.5  │ 0.5  │
╘═══════╧══════╧══════╛
CPD of M:
╒══════╤════════════════════╤═════════════════════╤════════════════════╤════════════════════╤════════════════════╤════════════════════╤═════════════════════╤════════════════════╤═════════════════════╤════════════════════╤═══════╤════════════════════╤════════════════════╤════════════════════╤════════════════════╤════════════════════╕
│ B    │ B(0)               │ B(0)                │ B(0)               │ B(0)

Focus your attention on the M CPD. In terms of extreme values, how does the table computed with the Bayesian method compare to the table computed with MLE? 

ANS: Bayesian model's CPD is quite similar to MLE method's. Because the graph with the strong 

prior and when the parameter is close to pesudo counts, it is closer to piror.

Next verify that the independencies of all the variables in your model are correct using the `local_independencies` method. Create and execute the code in the cell below to display the independencies in the CPD. 

In [6]:
bayes_model.local_independencies(['B','C','BW','CW','M'])

(B _|_ CW, C)
(C _|_ BW, B)
(BW _|_ CW, C | B)
(CW _|_ BW, B | C)

**Question:** Is your graph an I-map of the distribution and why?

ANS: Yes. This graph is an I-map because our distribution's the independence is

$$p(B,C,BW,CW,M) = p(B)\ p(C)\ p(BW\ |\ B)\ p(CW\ |\ C)\ p(M\ |\ BW,CW, C, B)$$  

And our graph's independeces are 

(B _|_ C, CW)
(C _|_ BW, B)
(BW _|_ CW,C | B)
(CW _|_ BW, B | C)

It could represents the independencies of the distribution.


Now you are ready to perform inference on your model. Use the belief propagation method to query the M variable with evidence that the cook is not the murderer. 

In [14]:
bayes_belief = BeliefPropagation(bayes_model)
bayes_belief_query = bayes_belief.query(variables=['M'], evidence={'C':0})
print(bayes_belief_query['M'])
bayes_belief.factors

╒═════╤══════════╕
│ M   │   phi(M) │
╞═════╪══════════╡
│ M_0 │   0.4192 │
├─────┼──────────┤
│ M_1 │   0.3095 │
├─────┼──────────┤
│ M_2 │   0.2713 │
╘═════╧══════════╛


C:\Users\genie\Anaconda3\lib\site-packages\pgmpy\factors\discrete\DiscreteFactor.py:586: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  phi.values = phi.values[slice_]
C:\Users\genie\Anaconda3\lib\site-packages\pgmpy\factors\discrete\DiscreteFactor.py:598: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  phi1.values = phi1.values[slice_]


defaultdict(list,
            {'B': [<DiscreteFactor representing phi(B:2) at 0x1568dafb630>,
              <DiscreteFactor representing phi(BW:2, B:2) at 0x1568dafb6d8>,
              <DiscreteFactor representing phi(M:3, B:2, BW:2, C:2, CW:2) at 0x1568dafb748>],
             'BW': [<DiscreteFactor representing phi(BW:2, B:2) at 0x1568dafb6d8>,
              <DiscreteFactor representing phi(M:3, B:2, BW:2, C:2, CW:2) at 0x1568dafb748>],
             'C': [<DiscreteFactor representing phi(C:2) at 0x1568dafb6a0>,
              <DiscreteFactor representing phi(CW:2, C:2) at 0x1568dafb710>,
              <DiscreteFactor representing phi(M:3, B:2, BW:2, C:2, CW:2) at 0x1568dafb748>],
             'CW': [<DiscreteFactor representing phi(CW:2, C:2) at 0x1568dafb710>,
              <DiscreteFactor representing phi(M:3, B:2, BW:2, C:2, CW:2) at 0x1568dafb748>],
             'M': [<DiscreteFactor representing phi(M:3, B:2, BW:2, C:2, CW:2) at 0x1568dafb748>]})

Compare this resulting marginal distribution to the marginal distribution you obtained for the same query in the previous homework using the CDP tables provided. How do these distribution differ?

ANS: The previous result is M_0 = 0.5520 , M_1 = 0.3680, M_2 = 0.0800.

It seems like M_2 is bigger and others are smaller because it is affected by pseudo distribtion and prior distribution.


Next, you will estimate the CPD parameters using the Bayesian estimator with a moderately weak but biased prior distribution. For this first estimate use the following moderately weak and uniform prior distributions (pseudo counts):

- C: {2,1}
- B: {1,2}
- CW: {1,2}
- BW: {2,1}
- M: {2,1,2}

In the cell below create and execute the code to perform Bayes estimation and display the CPD parameters. 

In [16]:
new_pseudo_counts = {'C': [2,1], 'B':[1,2], 'CW':[1,2], 'BW':[2,1], 'M':[2,1,2]}

new_bayes_model = BayesianModel([('B','BW'),('C','CW'),('B','M'),('C','M'),('BW','M'),("CW",'M')])

new_bayes_model.fit(dats, estimator=BayesianEstimator, prior_type='dirichlet', 
                  pseudo_counts=new_pseudo_counts)

for cpd in new_bayes_model.get_cpds():
    print("CPD of {variable}:".format(variable=cpd.variable))
    print(cpd)

CPD of B:
╒══════╤═════════╕
│ B(0) │ 0.26087 │
├──────┼─────────┤
│ B(1) │ 0.73913 │
╘══════╧═════════╛
CPD of BW:
╒═══════╤════════════════════╤════════════════════╕
│ B     │ B(0)               │ B(1)               │
├───────┼────────────────────┼────────────────────┤
│ BW(0) │ 0.6666666666666666 │ 0.6666666666666666 │
├───────┼────────────────────┼────────────────────┤
│ BW(1) │ 0.3333333333333333 │ 0.3333333333333333 │
╘═══════╧════════════════════╧════════════════════╛
CPD of C:
╒══════╤══════════╕
│ C(0) │ 0.695652 │
├──────┼──────────┤
│ C(1) │ 0.304348 │
╘══════╧══════════╛
CPD of CW:
╒═══════╤════════════════════╤════════════════════╕
│ C     │ C(0)               │ C(1)               │
├───────┼────────────────────┼────────────────────┤
│ CW(0) │ 0.3333333333333333 │ 0.3333333333333333 │
├───────┼────────────────────┼────────────────────┤
│ CW(1) │ 0.6666666666666666 │ 0.6666666666666666 │
╘═══════╧════════════════════╧════════════════════╛
CPD of M:
╒══════╤═══════╤═════════

Compare the parameters in the M CPD table to the estimate with a uniform prior. How are these different, and is this expected given the change in prior?

ANS: It has a weaker prior distribution than previous one. 

Now, use the belief propagation method to query the M variable with evidence that the cook is not the murderer.  

In [19]:
new_bayes_belief = BeliefPropagation(bayes_model)
new_bayes_belief_q = new_bayes_belief.query(variables=['M'], evidence={'C':0})
print(new_bayes_belief_q['M'])


╒═════╤══════════╕
│ M   │   phi(M) │
╞═════╪══════════╡
│ M_0 │   0.4192 │
├─────┼──────────┤
│ M_1 │   0.3095 │
├─────┼──────────┤
│ M_2 │   0.2713 │
╘═════╧══════════╛


C:\Users\genie\Anaconda3\lib\site-packages\pgmpy\factors\discrete\DiscreteFactor.py:586: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  phi.values = phi.values[slice_]
C:\Users\genie\Anaconda3\lib\site-packages\pgmpy\factors\discrete\DiscreteFactor.py:598: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  phi1.values = phi1.values[slice_]


Compare this marginal distribution to the one obtained with the a uniform prior. In general, are the differences what you would expect, and why?

ANS:  Pretty much. Because we have known that this distribution has different pseudo value.

Therefore, it has relatively weak prior.

## Part 2

Now you will explore how well the structure of the graph can be estimated. **Keep in mind that the graph used for the simulation is not a tree**. Further, your estimates of graph structure may differ from results obtained in other environment. Answer the questions based on the results you find. 

As a first step, execute the code i the cell below to create a simulated dataset with 20,000 cases. 

In [20]:
## set the sample size
nsamp = 2000

## First the conditionally independent variables
nr.seed(22234)
B_samps = pd.DataFrame(nr.binomial(1, 0.8, nsamp), columns = ['B'])
nr.seed(2355)
C_samps = pd.DataFrame(nr.binomial(1, 0.3, nsamp), columns = ['C'])

## Two variables conditionally depenent on one other
probs = [0.5, 0.1]
nr.seed(2134)
bern = sim_bernoulli(probs, nsamp)
BW_samps = pd.DataFrame(selec_dist_1(bern, B_samps, [0,1]), columns = ['BW'])

probs = [0.5, 0.7]
nr.seed(22234)
bern = sim_bernoulli(probs)
CW_samps = pd.DataFrame(selec_dist_1(bern, C_samps, [0,1]), columns = ['CW'])


probs = np.array([[0.0, 0.7, 0.5,  0.7, 0.0, 0.6, 0.5,  0.6, 0.0, 0.7, 0.5,  0.7, 0.0, 0.6, 0.5,  0.6],
                           [1.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0, 1.0, 0.0, 0.0,  0.0],
                           [0.0, 0.1, 0.4,  0.2, 0.0, 0.1, 0.3,  0.2, 0.0, 0.1, 0.4,  0.2, 0.0, 0.1, 0.3,  0.2]]).transpose()

nr.seed(2334)
sims = sim_multinoulli(probs, nsamp)

C_lg = [0,0,1,1,0,0,1,1,0,0,1,1,0,0,1,1]
B_lg = [0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1]
BW_lg = [0,0,0,0,1,1,1,1,0,0,0,0,1,1,1,1]
CW_lg = [0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1]
M_samps = pd.DataFrame(selec_dist_4(sims, B_samps, C_samps, BW_samps, CW_samps, B_lg, C_lg, BW_lg, CW_lg), columns = ['M'])



## Now concatenate the columns into one data frame
dats_big = pd.concat([B_samps, C_samps, BW_samps, CW_samps, M_samps], axis = 1)
print(dats_big.shape)
dats_big.head(20)

(2000, 5)


B  C  BW   CW    M
0   1  0   1  1.0  0.0
1   1  0   0  1.0  2.0
2   1  1   0  1.0  0.0
3   1  0   0  0.0  2.0
4   1  1   0  0.0  2.0
5   0  1   1  0.0  2.0
6   1  1   0  1.0  0.0
7   1  0   0  0.0  0.0
8   1  0   0  0.0  0.0
9   0  0   1  1.0  1.0
10  0  0   0  1.0  1.0
11  1  0   0  1.0  0.0
12  0  0   1  1.0  1.0
13  1  0   0  0.0  0.0
14  1  0   0  0.0  2.0
15  1  1   1  0.0  0.0
16  1  1   0  1.0  2.0
17  1  0   0  0.0  2.0
18  1  0   0  0.0  0.0
19  0  0   1  1.0  1.0

With the dataset simulated, you will now try estimating the model structure. Use the hill climb search algorithm along with the the BIC scoring function to estimate the model structure. Create and execute the code in the cell below to estimate the model structure and display the identified edges. 

In [34]:
from pgmpy.estimators import HillClimbSearch, BicScore, K2Score, StructureScore

est_bic = HillClimbSearch(dats_big, scoring_method=BicScore(dats_big))
bic_model = est_bic.estimate()
print(sorted(bic_model.nodes()))
BIC_edges = est_bic.estimate(max_indegree=2).edges()
BIC_edges

['B', 'BW', 'C', 'CW', 'M']


[('B', 'BW'), ('BW', 'CW'), ('BW', 'C'), ('CW', 'M')]

In [35]:
bic_model.local_independencies(bic_model.nodes())

(C _|_ CW, M, B | BW)
(CW _|_ C, B | BW)
(M _|_ C, BW, B | CW)

Examine these edges. Clearly, this structure is quite different from the structure used for the simulation. Keeping in mind the butler is much more likely to be the murderer compared to the cook, does this structure make sense and why?

ANS: It is hard to find the interpretation of this structure, because the local independencies of the

model is quite different compared to the previous model.

How good is this structure fit? To answer this question you will need to compare the BIC score of the graph used for the simulation with the BIC score of the estimated structure. You must create a baseline DAG structure and compare the BIC score to the score of the estimated model. 

In the cell below create the code to define a baseline DAG and compute and display the BIC score of both graphs.

In [33]:
bayes_model = BayesianModel([('B','BW'),('C','CW'),('B','M'),('C','M'),('BW','M'),("CW",'M')])

print(BicScore(dats_big).score(bayes_model))
print(BicScore(dats_big).score(bic_model))

-184.0713415026031
-56.62879646366832


Has the BIC score changed significantly? 

ANS: Yes. It has been changed a lot. -184 -> -56

Next, you will apply the K2 score method to the large dataset. In the cell below, create the code to use the hill climbing search with the K2 score to estimate and display the model structure. 

In [30]:
est_K2S = HillClimbSearch(dats_big, scoring_method=K2Score(dats_big))
K2S_model = est_K2S.estimate()
print(sorted(K2S_model.nodes()))
est_K2S_edges = est_K2S.estimate(max_indegree=2).edges()
est_K2S_edges

['B', 'BW', 'C', 'CW', 'M']


[('B', 'BW'), ('BW', 'CW'), ('BW', 'C'), ('CW', 'M')]

Is this graph structure any different from the one obtained with fewer cases?

ANS: The structure is the same. 

Now, compare the K2 score for the baseline model with the estimated model using the larger dataset for both cases. In the cell below create and execute the code to compute and display these scores. 

In [40]:
print(K2Score(dats_big).score(K2S_model))

-14.138593557466479


How are these K2 scores different? 

ANS: Yes, K2 is much more higher than before.

In the cell below create and execute the code to display the independencies of the graph structure you have found. 

In [39]:
K2S_model.local_independencies(K2S_model.nodes())

(C _|_ CW, M, B | BW)
(CW _|_ C, B | BW)
(M _|_ C, BW, B | CW)

Finally, compare the BIC and K2 scores of the two models you created with the K2 and BIC score methods on the large dataset. In the cell below create and execute the code to compute and display these 4 scores.  

In [41]:
print(BicScore(dats_big).score(K2S_model))
print(BicScore(dats_big).score(bic_model))
print(K2Score(dats_big).score(K2S_model))
print(K2Score(dats_big).score(bic_model))


-54.315024115936154
-54.315024115936154
-14.138593557466479
-14.138593557466479


Are there any substantial differences between the scores of these two models?

ANS: No, they are the same.